### DISCUSSION TITLE
Error while trying to install Halide

### DISCUSSION THREAD

**Discussion by Sofianebouaziz1** [halide/Halide#6727] [General]
Hello, I've been struggling a lot trying to install Halide from source since 2 days.

I have followed the README.md instructions as folllows :

git clone --depth 1 --branch llvmorg-12.0.0 https://github.com/llvm/llvm-project.git

cmake -DCMAKE_BUILD_TYPE=Release \
        -DLLVM_ENABLE_PROJECTS="clang;lld;clang-tools-extra" \
        -DLLVM_TARGETS_TO_BUILD="X86;ARM;NVPTX;AArch64;Mips;Hexagon;WebAssembly" \
        -DLLVM_ENABLE_TERMINFO=OFF -DLLVM_ENABLE_ASSERTIONS=ON \
        -DLLVM_ENABLE_EH=ON -DLLVM_ENABLE_RTTI=ON -DLLVM_BUILD_32_BITS=OFF \
        -S llvm-project/llvm -B llvm-buil

cmake --build llvm-build
cmake --install llvm-build --prefix llvm-install

export LLVM_ROOT=$PWD/llvm-install
export LLVM_CONFIG=$LLVM_ROOT/bin/llvm-config

git clone https://github.com/halide/Halide.git
cd Halide
make -j8

Everything worked fine ! Untill executing the make -j8 instruction and the error that is showing : 
../llvm-install/bin/clang file is not find

Thank you in advance !

**Comment by shoaibkamil** [comment]
I'm afraid it's going to be impossible to help without a full log.  Are you trying to build the main branch?  We dropped support for LLVM 12, but I'm not sure that's your issue.  Please post a full log, at least of the output after `make -j8`.

Have you tried using cmake instead of make to build Halide?

**Comment by alexreinking** [comment]
We do not support LLVM 12 anymore. See #6686 

Please try using CMake instead of Make to build Halide.

**Reply by Sofianebouaziz1** [reply]
Hello alex !

The reason I moved back to llvm12 because llvm13 caused the folowing error while executing :
cmake --build llvm-build

the error was in the following line : 
linking CXX executable ../../../../bin/clang  

is llvm14 supported in Halide ?

**Reply by alexreinking** [reply]
> is llvm14 supported in Halide?

The `main` branch of Halide supports the most recent release of LLVM, the previous release, and the `main` branch of LLVM.

**Reply by alexreinking** [reply]
> caused the folowing error while executing :

Again, we can't help you without full logs: the commands you ran, their output, the operating system you ran them on, the environment you ran them in, etc. etc.

**Reply by Sofianebouaziz1** [reply]
Thank you. I'll try installing llvm using the main branch.

For the logs I did not save them so I actually don't have them now. I'm using ubuntu 20.04.01. 

As I said I'll try installing llvm using the main branch and I'll make sure to save the logs this time.

**Comment by Sofianebouaziz1** [comment]
Hello again!

I will provide this time a full detail on my installation and where it didn't work.

I am using ubuntu 20.04.01, I've tried to install llvm using the main branch and the same error happened as the llvm 12.

So I will give you the instructions I've used + logs.


% git clone --depth 1 --branch main https://github.com/llvm/llvm-project.git

% cmake -DCMAKE_BUILD_TYPE=Release \
        -DLLVM_ENABLE_PROJECTS="clang;lld;clang-tools-extra" \
        -DLLVM_TARGETS_TO_BUILD="X86;ARM;NVPTX;AArch64;Mips;Hexagon;WebAssembly" \
        -DLLVM_ENABLE_TERMINFO=OFF -DLLVM_ENABLE_ASSERTIONS=ON \
        -DLLVM_ENABLE_EH=ON -DLLVM_ENABLE_RTTI=ON -DLLVM_BUILD_32_BITS=OFF \
        -S llvm-project/llvm -B llvm-build

% cmake --build llvm-build
% cmake --install llvm-build --prefix llvm-install

% export LLVM_ROOT=$PWD/llvm-install
% export LLVM_CONFIG=$LLVM_ROOT/bin/llvm-config

git clone --branch main https://github.com/halide/Halide.git
cd Halide
make -j8

Here is the logs file : 

[log.txt](https://github.com/halide/Halide/files/8559734/log.txt)


The error happens in the instruction : make -j8 (llvm-install/bin/clang is not found !).

**Reply by shoaibkamil** [reply]
Your logs seem to indicate that something was wrong with the way you configured LLVM.  It doesn't look like `clang` or `lld` were built when you built LLVM, which leads me to think the issue is how you initially ran `cmake`. To proceed, I would do the following:

1) `rm -rf llvm-build llvm-install Halide`
2) re-run cmake to set up your LLVM build:
```
$ cmake -DCMAKE_BUILD_TYPE=Release \
-DLLVM_ENABLE_PROJECTS="clang;lld;clang-tools-extra" \
-DLLVM_TARGETS_TO_BUILD="X86;ARM;NVPTX;AArch64;Mips;Hexagon;WebAssembly" \
-DLLVM_ENABLE_TERMINFO=OFF -DLLVM_ENABLE_ASSERTIONS=ON \
-DLLVM_ENABLE_EH=ON -DLLVM_ENABLE_RTTI=ON -DLLVM_BUILD_32_BITS=OFF \
-S llvm-project/llvm -B llvm-build
```
3) Build LLVM and install:
```
$ cmake --build llvm-build --parallel 8
$ cmake --install llvm-build --prefix llvm-install
```
4) Ensure that Clang was built and installed:
```
$ ls llvm-install/bin/clang
$ ./llvm-install/bin/clang --version
```
5) Continue the install steps, starting with setting the `LLVM_ROOT` environment variable

Note that the `\` are important when running CMake, otherwise it will not pick up the rest of the flags on the subsequent lines.  If I had to guess, I would hypothesize that's what went wrong here.

If you continue to have issues, please attach _complete_ build logs: what you attached before did not include the output of the initial `cmake` execution to configure your LLVM build, so I'm still guessing at what the issue is.

**Reply by Sofianebouaziz1** [reply]
Hello @shoaibkamil !
Thank you so much for your replay

I'll try building llvm again following the exact instructions you gave me. Thank you also for the --parallel flag that I didn't know about it seems like it speeds up the execution time since without it I was spending more than 10 hours to build llvm !

Have a great day !

**Reply by alexreinking** [reply]
>without it I was spending more than 10 hours to build llvm

I _strongly_ recommend installing Ninja with `sudo apt install ninja-build` and then adding `-G Ninja` to the CMake configure commands (the ones with the `-S` and `-B` flags, not the ones with `--build` or `--install`). Ninja is much more efficient about scheduling work on all of your cores than Make. If you do use Ninja, drop the `--parallel` flag to let Ninja determine the amount of parallelism to use automatically.

**Comment by Sofianebouaziz1** [comment]
I've tried both solutions using Ninja and --Parallel and ubuntu just freezes after every time I run them :/

**Comment by Sofianebouaziz1** [comment]
Hello again :( !
I tried to buil llvm again using the instrcution @shoaibkamil gave me (without the --parallel because it made my ubuntu freezes every time) and i still got an error.
Here is a detaillled log from when i started clonning to the error .
[log.txt](https://github.com/halide/Halide/files/8569737/log.txt)

**Reply by Sofianebouaziz1** [reply]
maybe that what caused --parallel to freeze

**Reply by Sofianebouaziz1** [reply]
using llvm 13 also gave the same error :
collect2: error: ld returned 1 exit status
make[2]: *** [tools/lto/CMakeFiles/LTO.dir/build.make:191 : lib/libLTO.so.13] Erreur 1
make[1]: *** [CMakeFiles/Makefile2:25555 : tools/lto/CMakeFiles/LTO.dir/all] Erreur 2
make: *** [Makefile:152 : all] Erreur 2

**Reply by alexreinking** [reply]
This is a failure in LLVM's build. You should report this to their project for assistance, here: https://github.com/llvm/llvm-project/issues/

**Comment by Sofianebouaziz1** [comment]
I fixed the problem, it was due to memory. The newer version of llvm takes a lot of memory so I downloaded llvm 11.0.1  .My question is : what is the branch of Halide 11 ?

**Reply by abadams** [reply]
The last version of Halide to support llvm 11 is old and frozen, so there's no point in compiling it yourself. You'd be better off just downloading the latest release.

**Comment by alexreinking** [comment]
These days, you should just try to use the `pip` package first 🙂 See the `README.md` for details.
